### PR TITLE
Feat persist

### PR DESCRIPTION
**Pull request by Taylor-lagrange**
[design documentation](https://yo93p0jbnb.feishu.cn/docx/doxcnRTiahtmEB3R4YAPe1A2kxc)

### CODE REVIEW

**Comment by Folyd** [comment]
Thanks for working on this. Did you run `cargo clippy` and `cargo fmt`?

**Comment by Taylor-lagrange** [comment]
> Thanks for working on this. Did you run `cargo clippy` and `cargo fmt`?

yes

**Review by Folyd — CHANGES_REQUESTED** [review]
Looks good overall. However, `clone()` is a litter bit too much, it's be better if the code more Rust idiomatic, such as iterating and error handling.

File: duo/src/data/persist.rs [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+                    Some(PersistMessage::Process(process)) => {
+                        if (!process.id.is_empty()) && (!process.service_name.is_empty()) {
+                            if let Err(e) = process_writer.write(ProcessPersist::from(process)).await {
+                                debug!("persist process info error: {:?}", e);
+                            }
+                        }
+                    }
+                    Some(PersistMessage::Data(data)) => {
+                        if !data.logs.is_empty() {
+                            for log in data.logs {
```
[the remark this answers is not in the document]
Comment (Taylor-lagrange) [reply duo/src/data/persist.rs]: If you use iterators, you need to use async closures, you need to capture trace_writer&log_writer, the ownership problem is difficult to solve.

[the remark this answers is not in the document]
Comment (Taylor-lagrange) [reply duo/src/data/persist.rs]: Other for loops have been modified（those for loops that don't have ownership issues）

File: duo/src/data/reader.rs [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+            let len = Self::get_len(&buffer[current_index..current_index + 4]);
+            current_index += 4;
+            let data: T = bincode::deserialize(&buffer[current_index..current_index + len]).unwrap();
+            data_vec.push(data);
+            current_index += len;
+        }
+        Ok(())
+    }
+
+    pub fn get_len(len: &[u8]) -> usize {
```
[the remark this answers is not in the document]
Comment (Taylor-lagrange) [reply duo/src/data/reader.rs]: function name has been modified and annotated for more unify semantics

**Comment by Taylor-lagrange** [comment]
> Thanks for working on this. Did you run `cargo clippy` and `cargo fmt`?

done

**Comment by Folyd** [comment]
Looks good. Thanks for working on this. @Taylor-lagrange

**Pull request merged**